llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -2755,17 +2755,22 @@ void InnerLoopVectorizer::fixupIVUsers(PHINode *OrigPhi,
       if (isa_and_nonnull<FPMathOperator>(II.getInductionBinOp()))
         B.setFastMathFlags(II.getInductionBinOp()->getFastMathFlags());
 
-      Value *CountMinusOne = B.CreateSub(
-          VectorTripCount, ConstantInt::get(VectorTripCount->getType(), 1));
-      CountMinusOne->setName("cmo");
-
       VPValue *StepVPV = Plan.getSCEVExpansion(II.getStep());
       assert(StepVPV && "step must have been expanded during VPlan execution");
       Value *Step = StepVPV->isLiveIn() ? StepVPV->getLiveInIRValue()
                                         : State.get(StepVPV, VPLane(0));
-      Value *Escape =
-          emitTransformedIndex(B, CountMinusOne, II.getStartValue(), Step,
-                               II.getKind(), II.getInductionBinOp());
+      Value *Escape = nullptr;
+      if (EndValue->getType()->isIntegerTy())
+        Escape = B.CreateSub(EndValue, Step);
```

Comment:
Is this guaranteed to be the same value as before? For example, do we know for sure that the overflow behaviour in each scenario is the same?

Basically does this statement hold true:

`start + (n - 1)` == `end - 1`

where I assume that `end` is equivalent to `start + n`. For example, I can imagine a scenario where `start + (n - 1)` is UINT64_MAX and so `end` might be 0. Perhaps it doesn't matter since `end` will underflow back to UINT64_MAX?

Reply:
Yes it should always produce the same result, except if the step is `undef` as each use of the `undef` value may see a different value: https://alive2.llvm.org/ce/z/DS8Cuc. 

But I don't think the `undef` case is relevant here; if `step` would be `undef`, then `(tc - 1) * step` could also produce a different result to the scalar induction incremented by `undef` in each loop iteration.